ci/src/Registry/Index.purs

```
@@ -0,0 +1,128 @@
+module Registry.Index where
+
+import Registry.Prelude
+
+import Control.Alternative (guard)
+import Data.Argonaut (decodeJson, encodeJson, parseJson, stringify)
+import Data.Array as Array
+import Data.Array.NonEmpty (NonEmptyArray)
+import Data.Array.NonEmpty as NEA
+import Data.Map as Map
+import Data.String as String
+import Data.String.Pattern (Pattern(..))
+import Foreign.SemVer (SemVer)
+import Node.FS.Aff (readTextFile, writeTextFile)
+import Node.Glob.Basic (expandGlobsCwd)
+import Partial.Unsafe (unsafeCrashWith)
+import Registry.PackageName (PackageName(..))
+import Registry.Schema (Manifest)
+
+type RegistryIndex = Map PackageName (Map SemVer Manifest)
+
+-- This function must be run from the root of the registry index.
+-- NOTE: Right now, this assumes that manifest files will parse
+readRegistryIndex :: Aff RegistryIndex
+readRegistryIndex = do
+  packagePaths <- Array.fromFoldable <$> expandGlobsCwd [ "*" ]
+  let
+    -- Exclude certain files that will always be in the root of the registry index.
+    -- These files do not correspond to a package manifest file.
+    exclude :: Array String
+    exclude = [ "config.json" ]
+
+    goPath path = do
+      fileName <- Array.last $ String.split (Pattern "/") path
+      guard (not (Array.elem fileName exclude))
+      pure $ PackageName fileName
+
+    packages = Array.mapMaybe goPath packagePaths
+
+  parsed <- for packages \package -> Tuple package <$> readPackage package
+  let
+    parsedPackages :: Array (Tuple PackageName (NonEmptyArray Manifest))
+    parsedPackages = Array.mapMaybe (\(Tuple package mbManifests) -> Tuple package <$> mbManifests) parsed
```

Here we're ignoring files that don't parse, but possibly we should throw an exception so that it's easier for us to find out when (for some should-not-be-possible reason) a manifest we've written doesn't parse.

---

Thoughts, @f-f?

---

Totally. Also I'd go as far as including a CI check that validates all the files against the `Manifest` type, like we [do here already](https://github.com/purescript/registry/blob/master/ci/scripts/verify-dhall.sh). We don't have to include this CI check in this PR though (especially because the index is not there yet)

---

Good idea! I'd say skip the CI check in this PR, @colinwahl, but we can switch to throwing an exception here.